lib/elixir/lib/calendar/duration.ex

```diff
@@ -5,10 +5,77 @@ defmodule Duration do
 
-  Date and time scale units are represented as integers, allowing for both positive and negative values.
+  Date and time scale units are represented as integers, allowing for
+  both positive and negative values.
 
-  Microseconds are represented using a tuple `{microsecond, precision}`. This ensures compatibility with
-  other calendar types implementing time, such as `Time`, `DateTime`, and `NaiveDateTime`.
+  Microseconds are represented using a tuple `{microsecond, precision}`.
+  This ensures compatibility with other calendar types implementing time,
+  such as `Time`, `DateTime`, and `NaiveDateTime`.
+
+  ## Shifting
+
+  The most common use of durations in Elixir's standard library is to
+  "shift" the calendar types.
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      ~D[2016-03-03]
+
+  In the example above, `Date.shift/2` automatically converts the units
+  into a Duration struct, although one can also be given directly:
+
+      iex> Date.shift(~D[2016-01-03], Duration.new!(month: 2))
+      ~D[2016-03-03]
+
+  It is important to notice that shifting is not an arithmetic operation
+  For example, adding `date + 1 month + 1 month` does not yield the same
+  result as `date + 2 months`. Let's see an example:
+
+      iex> ~D[2016-01-31] |> Date.shift(month: 1) |> Date.shift(month: 1)
+      ~D[2016-03-29]
+
+      iex> ~D[2016-01-31] |> Date.shift(month: 2)
+      ~D[2016-03-31]
+
+  As you can see above, the results differ, which explains why operations
+  with durations are called "shift" rather than "add".
+
+  In particular, when applying durations to Calendar.ISO types:
+
+    * larger units (such as years and months) are applied before
+      smaller ones (such as weeks, hours, days, etc)
+
+    * in case of non-existing dates, the results are rounded down to the
+      nearest valid date
+
+  ## Intervals
+
+  Durations in Elixir can be combined with stream operations to build intervals.
+  For example, to retrieve the next three wednesdays starting from 17th April, 2024:
+
+      iex> ~D[2024-04-17] |> Stream.iterate(&Date.shift(&1, week: 1)) |> Enum.take(3)
+      [~D[2024-04-17], ~D[2024-04-24], ~D[2024-05-01]]
+
+  However, once again, it is important to remember that shifting a duration is not
+  arithmetic, so you may want to use the functions in this module depending on what
+  you to achieve. Compare the results of both examples below:
+
+      # Adding one month after the other
+      iex> date = ~D[2016-01-31]
+      iex> duration = Duration.new!(month: 1)
+      iex> stream = Stream.iterate(date, fn prev_date -> Date.shift(prev_date, duration) end)
+      iex> Enum.take(stream, 3)
+      [~D[2016-01-31], ~D[2016-02-29], ~D[2016-03-29]]
+
+      # Multiplying durations one after the other
+      iex> date = ~D[2016-01-31]
+      iex> duration = Duration.new!(month: 1)
+      iex> stream = Stream.counter(fn i -> Date.shift(date, Duration.multiply(duration, i)) end)
```

Comment:
Is it weird to call it a counter when it doesn't count up numbers per se? Is there a better name?

Reply:
Maybe `Stream.from_index` to pair with `Stream.with_index`?

Reply:
Yeah, I'd search for a better name too. In the meantime here's a, if I may say so, nice usage of this feature, a simple exponential backoff:

```elixir
iex> Stream.counter(&Integer.pow(2, &1) * 1000) |> Enum.take(5)
[1000, 2000, 4000, 8000, 16000]
```

Reply:
I understand the infinite stream implementation can be more generic, but I wonder if just using Enum.map on a range wouldn’t be simpler to illustrate how to do this ?

Reply:
Also, maybe an infinite range with just a start and a step could offer more flexibility?
Example from python:
https://docs.python.org/3/library/itertools.html#itertools.count

Reply:
@sabiwara the range and step is pretty much `Stream.iterate/1`. Well, the whole thing could be done with iterate+map indeed, so yeah... and all of them could be done with `unfold/1` anyway.

In any case, I have renamed to `from_index` because it feels like less of a departure from the existing API.